File: src/main/java/com/github/dylmeadows/eontimer/core/timer/Gen4TimerFactory.java
package com.github.dylmeadows.eontimer.core.timer;

import com.github.dylmeadows.eontimer.model.Stage;
import com.github.dylmeadows.eontimer.model.config.TimerConfigurationModel;
import com.github.dylmeadows.eontimer.model.timer.Gen4TimerMode;
import com.github.dylmeadows.eontimer.model.timer.Gen4TimerModel;
import com.github.dylmeadows.eontimer.util.CalibrationUtils;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

@Component
@RequiredArgsConstructor(onConstructor = @__(@Autowired))
public class Gen4TimerFactory implements TimerFactory {

    private final Gen4TimerModel timerModel;
    private final TimerConfigurationModel timerConfig;

    @Override
    public List<Stage> createTimer() {
        if (timerModel.getMode() == Gen4TimerMode.STANDARD) {
            return getStages();
        }
        return Collections.emptyList();
    }

    private List<Stage> getStages() {
        List<Integer> stages = new ArrayList<>();
        int calibration = CalibrationUtils.createCalibration(
            timerModel.getCalibratedDelay(),
            timerModel.getCalibratedSecond(),
            timerConfig.getConsole());
        stages.add(normalize(normalize(timerModel.getTargetSecond() * 1000 + calibration + 200)
            - CalibrationUtils.convertToMillis(timerModel.getTargetDelay(), timerConfig.getConsole())));
        stages.add(CalibrationUtils.convertToMillis(timerModel.getTargetDelay(), timerConfig.getConsole()) - calibration);
        return stages.stream()
            .map(Stage::new)
            .collect(Collectors.toList());
    }

    private int normalize(int stage) {
        while (stage < timerConfig.getMinimumLength())
            stage += 60000;
        return stage;
    }
}

File: src/main/kotlin/io/github/dylmeadows/eontimer/service/factory/Gen4TimerFactory.kt
package io.github.dylmeadows.eontimer.service.factory

import io.github.dylmeadows.eontimer.model.TimerState
import io.github.dylmeadows.eontimer.model.timer.Gen4TimerModel
import io.github.dylmeadows.eontimer.service.CalibrationService
import io.github.dylmeadows.eontimer.service.factory.timer.DelayTimerFactory
import io.github.dylmeadows.eontimer.util.asFlux
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.time.Duration
import javax.annotation.PostConstruct

@Component
class Gen4TimerFactory @Autowired constructor(
    private val timerState: TimerState,
    private val gen4TimerModel: Gen4TimerModel,
    private val delayTimerFactory: DelayTimerFactory,
    private val calibrationService: CalibrationService) : TimerFactory {

    private val calibration: Long
        get() = calibrationService.createCalibration(gen4TimerModel.calibratedDelay, gen4TimerModel.calibratedSecond)

    @PostConstruct
    private fun initialize() {
        listOf(
            gen4TimerModel.calibratedDelayProperty,
            gen4TimerModel.calibratedSecondProperty,
            gen4TimerModel.targetDelayProperty,
            gen4TimerModel.targetSecondProperty)
            .map { it.asFlux() }
            .forEach {
                it.subscribe {
                    timerState.update(stages)
                }
            }
    }

    override val stages: List<Duration>
        get() = delayTimerFactory.createStages(
            gen4TimerModel.targetSecond,
            gen4TimerModel.targetDelay,
            calibration)

    override fun calibrate() {
        gen4TimerModel.calibratedDelay +=
            calibrationService.calibrateToDelays(
                delayTimerFactory.calibrate(
                    gen4TimerModel.targetDelay,
                    gen4TimerModel.delayHit))
    }
}